---
author: "University of Melbourne Centre for Cancer Research"
date: "`r format(Sys.time(), '%a %Y-%b-%d %H:%M')`"
output:
  html_document:
    code: hide
    theme: cosmo
    toc: true
    code_download: true
  rmdformats::material:
    highlight: kate
params:
  sample: "SAMPLE"
  snv_vcf: "/path/to/snv.vcf"
  sv_vcf: "/path/to/sv.vcf"
  cnv_tsv: "/path/to/purple.somatic.cnv"
  snvoutdir: "/path/to/outdir_snv"
title: "`r glue::glue('HRD Results from CHORD & HRDetect')`"
description: "HRD Results"
---

<style type="text/css">
.main-container {
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r render_report_interactively, eval=F, echo=F}
params <- function(x) {
  paths <- list(
    local = list(
      sample = "SAMPLE",
      snv_vcf = here::here("inst/extdata/umccrise/v0.18/snv/somatic-ensemble-PASS.vcf.gz"),
      sv_vcf = here::here("inst/extdata/umccrise/v0.18/sv/manta.vcf.gz"),
      cnv_tsv = here::here("inst/extdata/purple/v2.39/purple.cnv.somatic.tsv"),
      snvoutdir = here::here("nogit/hrdetect/out/snvoutdir")
      
    )
  )
    paths[[x]]
}
params <- params("local")

render_me <- function() {
  rmarkdown::render(
    input = here::here("inst/rmd/wgs/hrd_detection/hrdetect/hrdetect.Rmd"),
    params = params)
}
render_me()
```

```{r load_pkgs, message=F, warning=F}
require(BSgenome.Hsapiens.UCSC.hg38)
require(BSgenome.Hsapiens.1000genomes.hs37d5)
require(CHORD)
require(DT)
require(gpgr)
require(gt)
require(here)
require(kableExtra)
require(mutSigExtractor)
require(signature.tools.lib)
require(tidyverse)
```

## Introduction

Here we are using
[CHORD](https://github.com/UMCUGenetics/CHORD) and
[HRDetect](https://github.com/Nik-Zainal-Group/signature.tools.lib)
to infer possible HRD deficiency in several of our in-house samples.
Input files used are:

* SNVs/INDELs: filtered umccrised `somatic-ensemble-PASS` VCF files.
* SVs: filtered umccrised `manta` VCF files.
* CNVs: PURPLE somatic CNV files.

The samples used for HRD predictions are shown below, along with UMCCR's comments
for HRD likelihood.
Note that HRDetect doesn't work with GRCh37, so only samples with hg38 data were
used with that tool.

## Samples

```{r setup_data, message=F, warning=F}
s <- tibble::tribble(
  ~nm, ~prefix, ~genome, ~comment,
  "P033", "2016_249_17_MH_P033__CCR170115b_MH17T002P033", "hg38", "ValidationSample",
  "P025", "2016_249_18_WH_P025__CCR180149_VPT-WH025-E", "hg38", "ValidationSample",
  "BALL10", "B_ALL_Case_10__MDX190025_B_ALL_Case_10T", "hg38", "ValidationSample",
  "CUP8", "CUP-Pairs8__PRJ180660_8_DNA009529_FFPE", "hg38", "ValidationSample",
  "SEQCII50", "SEQC-II-50pc__SEQC-II_Tumor_50pc", "hg38", "ValidationSample",
  "SFRC01073", "SFRC01073__PRJ180598_SFRC01073-S2T", "hg38", "ValidationSample",
  "SFRC01178", "SBJ00267__SBJ00267_PRJ200047_L2000059", "hg38", "Unlikely",
  "SFRC01122", "SBJ00001__SBJ00001_PRJ190733_L1900931", "hg38", "Likely",
  "SFRC01143", "SFRC01143__PRJ190190_SFRC01143_T", "GRCh37", "Likely",
  "SFRC01160", "SFRC01160__PRJ190376_SFRC01160-S1-19-1434", "GRCh37", "Likely",
  "SFRC01210", "SBJ00162__MDX190151", "GRCh37", "Likely",
  "PM3056445", "PM3056445__MDX190101_DNA052297-T", "GRCh37", "Likely",
  "PM3028142", "SBJ00286__SBJ00286_PRJ200069_L2000117", "hg38", "Likely",
  "PM4024666", "SBJ00322__SBJ00322_MDX200054_L2000238", "hg38", "Unlikely",
  "PM3061034", "SBJ00301__SBJ00301_MDX200046_L2000147", "hg38", "Unlikely",
  "SFRC01258", "SBJ00281__SBJ00281_PRJ200061_L2000100", "hg38", "Unlikely",
  "PM2293366", "SBJ00236__SBJ00236_MDX190218_L1901019", "hg38", "Unlikely",
  "SFRC01180", "SBJ00303__SBJ00303_PRJ200200_L2000150", "hg38", "Likely",
  ) %>%
  dplyr::distinct() %>%
  dplyr::mutate(
    snvindel_vcf = file.path(here::here("nogit/inputs/snv"), paste0(prefix, "-somatic-ensemble-PASS.vcf.gz")),
    sv_vcf = file.path(here::here("nogit/inputs/sv"), paste0(prefix, "-manta.vcf.gz")),
    cnv_file = file.path(here::here("nogit/inputs/cnv"), paste0(prefix, ".purple.cnv")),
    snvoutdir = file.path(here::here("nogit/hrdetect/snvoutdir"), prefix)
  )

# HRDetect doesn't work with GRCh37
s_hg38 <- s %>%
  dplyr::filter(genome == "hg38")

```

```{r inputs_table}
s %>%
  dplyr::select(Sample = nm, UMCCR_Comment = comment, Assembly = genome, Prefix = prefix) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left")
```

## Run CHORD

```{r run_chord, eval=FALSE}
res_chord <- seq_len(nrow(s)) %>%
  purrr::map(function(i) {
    cat(s$nm[i])
    gpgr::run_chord(snv = s$snvindel_vcf[i],
                    sv = s$sv_vcf[i],
                    sample = s$nm[i],
                    genome = s$genome[i])
  })
saveRDS(res_chord, file = here::here("nogit/chord/res_2020-05-20.rds"))
```

```{r process_chord_results}
res_chord <- readRDS(here::here("nogit/chord/res_2020-05-20.rds"))

sigs <- res_chord %>%
  purrr::map("contexts") %>%
  do.call(rbind, .) %>%
  tibble::as_tibble(rownames = "sample")

sigs_transpose <- sigs %>%
  tidyr::pivot_longer(-sample, names_to = "context", values_to = "count") %>%
  tidyr::pivot_wider(names_from = sample, values_from = count)

CHORD_hrd_preds <- res_chord %>%
  purrr::map("prediction") %>%
  dplyr::bind_rows()
```

## CHORD Results {.tabset .tabset-fade}

### HRD Predictions
```{r chord_tables_hrdpreds}
CHORD_hrd_preds <- CHORD_hrd_preds %>%
  dplyr::left_join(
    s %>% dplyr::select(sample = nm, UMCCR_Comment = comment),
    by = "sample"
  ) %>%
  dplyr::relocate(UMCCR_Comment, hr_status, "p_hrd.50%", .after = sample) %>%
  dplyr::rename(CHORD_HRD = hr_status, "CHORD_prob_50%" = "p_hrd.50%")

CHORD_hrd_preds %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "HRD predictions.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 300, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip')) %>%
  DT::formatStyle(
    "CHORD_HRD",
    backgroundColor = DT::styleEqual(
      c("HR_deficient", "HR_proficient"), c("lightblue", "lightpink")
    )
  ) %>%
  DT::formatStyle(
    "UMCCR_Comment",
    backgroundColor = DT::styleEqual(
      c("Likely", "Unlikely"), c("lightblue", "lightpink")
    )
  )
```

### Sig Counts

```{r chord_tables_sigs}
sigs_transpose %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "Counts of mutation context per sample.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 300, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip'))
```

## Run HRDetect

```{r hrdetect_run, eval=FALSE}
res_hrdetect <- seq_len(nrow(s_hg38)) %>%
  purrr::map(function(i) {
    cat(s_hg38$nm[i], "\n")
    gpgr::hrdetect_run(
      nm = s_hg38$nm[i],
      snvindel_vcf = s_hg38$snvindel_vcf[i],
      sv_vcf = s_hg38$sv_vcf[i],
      cnv_file = s_hg38$cnv_file[i],
      genome = s_hg38$genome[i],
      snvoutdir = s_hg38$snvoutdir[i])
  })
res_hrdetect <- dplyr::bind_rows(res_hrdetect)
saveRDS(res_hrdetect, file = here::here("nogit/hrdetect/res_2020-05-20.rds"))
```

## HRDetect Results

```{r hrdetect_results}
res_hrdetect <- readRDS(here::here("nogit/hrdetect/res_2020-05-20.rds"))
res_hrdetect %>%
  gt(rowname_col = "sample") %>%
  fmt_percent(columns = "Probability") %>%
  fmt_number(columns = c("intercept", "del.mh.prop", "SNV3", "SV3", "SV5", "hrd", "SNV8"), decimals = 2)
```


## CHORD & HRDetect Table

```{r}
CHORD_hrd_preds %>%
  dplyr::left_join(res_hrdetect, by = "sample") %>%
  dplyr::relocate(Probability, .after = "CHORD_HRD") %>%
  dplyr::rename(HRDetect_probability = Probability) %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "HRD predictions.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 600, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip')) %>%
  DT::formatStyle(
    "CHORD_HRD",
    backgroundColor = DT::styleEqual(
      c("HR_deficient", "HR_proficient"), c("lightblue", "lightpink")
    )
  ) %>%
  DT::formatStyle(
    "UMCCR_Comment",
    backgroundColor = DT::styleEqual(
      c("Likely", "Unlikely"), c("lightblue", "lightpink")
    )
  ) %>%
    DT::formatPercentage(
    c("HRDetect_probability", "CHORD_prob_50%"),
    digits = 1
  ) %>%
  DT::formatStyle(
    c("HRDetect_probability", "CHORD_prob_50%"),
    backgroundColor = styleInterval(0.1, c('lightpink', 'lightblue'))
  )
```

