---
author: "University of Melbourne Centre for Cancer Research"
date: "`r format(Sys.time(), '%a %Y-%b-%d %H:%M')`"
output:
  html_document:
    code: hide
    theme: cosmo
    toc: true
    code_download: true
  rmdformats::material:
    highlight: kate
params:
  project: "apgi"
title: "`r glue::glue('HRD Results from CHORD & HRDetect')`"
description: "HRD Results"
---

<style type="text/css">
.main-container {
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r render_report_interactively, eval=F, echo=F}
params <- function(x) {
  paths <- list(
    local = list(
      project = "apgi"
    )
  )
    paths[[x]]
}
params <- params("local")

render_me <- function() {
  rmarkdown::render(
    input = here::here("inst/rmd/wgs/hrd_detection/chord_and_hrdetect_apgi.Rmd"),
    params = params)
}
render_me()
```

```{r load_pkgs, message=F, warning=F}
require(BSgenome.Hsapiens.UCSC.hg38)
require(BSgenome.Hsapiens.1000genomes.hs37d5)
require(CHORD)
require(DT)
require(glue)
require(gpgr)
require(gt)
require(here)
require(kableExtra)
require(mutSigExtractor)
require(signature.tools.lib)
require(tidyverse)
```

## Introduction

Here we are using
[CHORD](https://github.com/UMCUGenetics/CHORD) and
[HRDetect](https://github.com/Nik-Zainal-Group/signature.tools.lib)
to infer possible HRD deficiency in several of our in-house samples.
Input files used are:

* SNVs/INDELs: filtered umccrised `somatic-ensemble-PASS` VCF files.
* SVs: filtered umccrised `manta` VCF files.
* CNVs: PURPLE somatic CNV files.

The samples used for HRD predictions are shown below, along with UMCCR's comments
for HRD likelihood.
Note that HRDetect doesn't work with GRCh37, so only samples with hg38 data were
used with that tool.

## Samples

```{r setup_data, message=F, warning=F}
project <- params$project

apgi2icgc <-
  here::here("nogit/inputs/apgi/apgi2icgc.tsv") %>%
  readr::read_tsv(col_types = "cc")

# 'Truth'
icgc_hrd <-
  here::here("nogit/inputs/apgi/sean/HRDetect_and_MSIpositive_cases_in_AUS-PACA.csv") %>%
  readr::read_csv(col_types = readr::cols(.default = "c")) %>%
  dplyr::select(ICGC_ID, DNA_repair_status, MSI_Sensor_Score, HRDetect_probability) %>%
  dplyr::left_join(apgi2icgc, by = "ICGC_ID")

# Only one sample in the Truth set (ICGC_0006)
s <-
  here::here("nogit/inputs/apgi/cancer_reports") %>%
  list.files(pattern = "html") %>%
  tibble::tibble(x = .) %>%
  dplyr::mutate(nm = sub("__.*", "", x),
                prefix = sub("_cancer_report\\.html", "", x),
                genome = "hg38",
                comment = "UNSURE") %>%
  dplyr::distinct() %>%
  dplyr::mutate(
    snvindel_vcf = file.path(here::here("nogit/inputs/apgi/snv"), paste0(prefix, "-somatic-PASS.vcf.gz")),
    sv_vcf = file.path(here::here("nogit/inputs/apgi/sv"), paste0(prefix, "-manta.vcf.gz")),
    cnv_file = file.path(here::here("nogit/inputs/apgi/cnv"), paste0(prefix, ".purple.cnv.somatic.tsv")),
    snvoutdir = file.path(here::here("nogit/hrdetect/snvoutdir"), prefix)
  ) %>%
  dplyr::select(-x) %>%
  dplyr::mutate(APGI_ID = nm,
                APGI_ID = sub("LTS_", "", APGI_ID)) %>%
  dplyr::left_join(icgc_hrd, by = "APGI_ID")
```

```{r inputs_table}
s %>%
  dplyr::select(APGI_ID, ICGC_ID, Assembly = genome, Prefix = prefix) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left")
```

## Run CHORD

```{r run_chord, eval=FALSE}
res_chord <- seq_len(nrow(s)) %>%
  purrr::map(function(i) {
    cat(s$nm[i])
    x <- gpgr::run_chord(snv = s$snvindel_vcf[i],
                         sv = s$sv_vcf[i],
                         sample = s$nm[i],
                         genome = s$genome[i])
    saveRDS(x, file = here(glue("nogit/chord/rds/{project}/{s$prefix[i]}.rds")))
    x
  })
saveRDS(res_chord, file = here(glue("nogit/chord/rds/{project}/all_2020-05-28.rds")))
```

```{r process_chord_results}
res_chord <- readRDS(file = here(glue("nogit/chord/rds/{project}/all_2020-05-28.rds")))

sigs <- res_chord %>%
  purrr::map("contexts") %>%
  do.call(rbind, .) %>%
  tibble::as_tibble(rownames = "sample")

sigs_transpose <- sigs %>%
  tidyr::pivot_longer(-sample, names_to = "context", values_to = "count") %>%
  tidyr::pivot_wider(names_from = sample, values_from = count)

CHORD_hrd_preds <- res_chord %>%
  purrr::map("prediction") %>%
  dplyr::bind_rows()
```

## CHORD Results {.tabset .tabset-fade}

### HRD Predictions
```{r chord_tables_hrdpreds}
CHORD_hrd_preds <- CHORD_hrd_preds %>%
  dplyr::left_join(
    s %>% dplyr::select(sample = nm, genome, UMCCR_Comment = comment),
    by = "sample"
  ) %>%
  dplyr::relocate(genome, UMCCR_Comment, hr_status, "p_hrd.50%", .after = sample) %>%
  dplyr::rename(CHORD_HRD = hr_status, "CHORD_prob_50%" = "p_hrd.50%")

CHORD_hrd_preds %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "HRD predictions.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 300, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip')) %>%
  DT::formatStyle(
    "CHORD_HRD",
    backgroundColor = DT::styleEqual(
      c("HR_deficient", "HR_proficient"), c("lightblue", "lightpink")
    )
  ) %>%
  DT::formatStyle(
    "UMCCR_Comment",
    backgroundColor = DT::styleEqual(
      c("Likely", "Unlikely"), c("lightblue", "lightpink")
    )
  )
```

### Sig Counts

```{r chord_tables_sigs}
sigs_transpose %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "Counts of mutation context per sample.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 300, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip'))
```

## Run HRDetect

```{r hrdetect_run, eval=FALSE}
res_hrdetect <- seq_len(nrow(s)) %>%
  purrr::map(function(i) {
    cat(s$nm[i], "\n")
    if (s$nm[i] == "SFRC01143") { # PURPLE without minoralleleploidy
      return(tibble::tribble(
        ~sample, ~Probability, ~intercept, ~del.mh.prop, ~SNV3, ~SV3, ~SV5, ~hrd, ~SNV8,
        s$nm[i], NA,           NA,         NA,           NA,    NA,   NA,   NA,   NA,
      ))

    }
    x <- gpgr::hrdetect_run(
      nm = s$nm[i],
      snvindel_vcf = s$snvindel_vcf[i],
      sv_vcf = s$sv_vcf[i],
      cnv_file = s$cnv_file[i],
      genome = s$genome[i],
      snvoutdir = s$snvoutdir[i])
    saveRDS(x, file = here(glue("nogit/hrdetect/rds/{project}/{s$prefix[i]}.rds")))
    x

  })
res_hrdetect <- dplyr::bind_rows(res_hrdetect)
saveRDS(res_hrdetect, file = here(glue("nogit/hrdetect/rds/{project}/all_2020-06-01.rds")))
```

## HRDetect Results

```{r hrdetect_results}
res_hrdetect <- readRDS(file = here(glue("nogit/hrdetect/rds/{project}/all_2020-06-01.rds")))
res_hrdetect %>%
  gt(rowname_col = "sample") %>%
  fmt_percent(columns = "Probability") %>%
  fmt_number(columns = c("intercept", "del.mh.prop", "SNV3", "SV3", "SV5", "hrd", "SNV8"), decimals = 2)
```


## CHORD & HRDetect Table

```{r chord_hrd_tab}
CHORD_hrd_preds %>%
  dplyr::left_join(res_hrdetect, by = "sample") %>%
  dplyr::relocate(Probability, .after = "CHORD_HRD") %>%
  dplyr::rename(HRDetect_probability = Probability) %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                caption = "HRD predictions.",
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 700, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip')) %>%
  DT::formatStyle(
    "CHORD_HRD",
    backgroundColor = DT::styleEqual(
      c("HR_deficient", "HR_proficient"), c("lightblue", "lightpink")
    )
  ) %>%
  DT::formatStyle(
    "UMCCR_Comment",
    backgroundColor = DT::styleEqual(
      c("Likely", "Unlikely"), c("lightblue", "lightpink")
    )
  ) %>%
    DT::formatPercentage(
    c("HRDetect_probability", "CHORD_prob_50%"),
    digits = 1
  ) %>%
  DT::formatStyle(
    c("HRDetect_probability", "CHORD_prob_50%"),
    backgroundColor = styleInterval(0.1, c('lightpink', 'lightblue'))
  )
```

