---
author: "University of Melbourne Centre for Cancer Research"
date: "`r format(Sys.time(), '%a %Y-%b-%d %H:%M')`"
output:
  html_document:
    code: hide
    theme: cosmo
    toc: false
    code_download: true
  rmdformats::material:
    highlight: kate
params:
  sample: "SAMPLE"
  sv: "/path/to/sv.vcf"
  snv: "/path/to/snv.vcf"
title: "`r glue::glue('{params$sample} HRDetect Results')`"
description: "Results from HRDetect analysis"
---

<style type="text/css">
.main-container {
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r render_report_interactively, eval=F, echo=F}
params <- function(x) {
  paths <- list(
    local = list(
      sample = "SAMPLE",
      sv = here::here("inst/extdata/umccrise/v0.18/sv/manta.vcf.gz"),
      snv = here::here("inst/extdata/umccrise/v0.18/snv/somatic-ensemble-PASS.vcf.gz")
    )
  )
    paths[[x]]
}
params <- params("local")

render_me <- function() {
  rmarkdown::render(
    input = here::here("inst/rmd/wgs/hrd_detection/hrdetect/hrdetect.Rmd"),
    params = params)
}
render_me()
```

```{r load_pkgs, message=F, warning=F}
require(here)
require(signature.tools.lib)
require(tidyverse)
```



## Set up variables

```{r var_setup}
sample_names <- c("sample1", "sample2")
SNV_tab_files <- c(here("nogit/hrdetect/testthat/test_hrdetect_1/test_hrdetect_1.snv.simple.txt"),
                   here("nogit/hrdetect/testthat/test_hrdetect_2/test_hrdetect_2.snv.simple.txt"))
SV_bedpe_files <-   c(here("nogit/hrdetect/testthat/test_hrdetect_1/test_hrdetect_1.sv.bedpe"),
                      here("nogit/hrdetect/testthat/test_hrdetect_2/test_hrdetect_2.sv.bedpe"))
Indels_vcf_files <- c(here("nogit/hrdetect/testthat/test_hrdetect_1/test_hrdetect_1.indel.vcf.gz"),
                      here("nogit/hrdetect/testthat/test_hrdetect_2/test_hrdetect_2.indel.vcf.gz"))
CNV_tab_files <-    c(here("nogit/hrdetect/testthat/test_hrdetect_1/test_hrdetect_1.cna.txt"),
                      here("nogit/hrdetect/testthat/test_hrdetect_2/test_hrdetect_2.cna.txt"))

# name the vector entries with the sample names
names(SNV_tab_files) <- sample_names
names(SV_bedpe_files) <- sample_names
names(Indels_vcf_files) <- sample_names
names(CNV_tab_files) <- sample_names
```


```{r}
# load SNV data and convert to SNV mutational catalogues
SNVcat_list <- list()
for (i in 1:length(SNV_tab_files)){
  tmpSNVtab <- read.table(SNV_tab_files[i], sep = "\t",
                          header = TRUE, check.names = FALSE,
                          stringsAsFactors = FALSE)
  # convert to SNV catalogue, see ?tabToSNVcatalogue or
  # ?vcfToSNVcatalogue for details
  res <- tabToSNVcatalogue(subs = tmpSNVtab, genome.v = "hg19")
  colnames(res$catalogue) <- sample_names[i]
  SNVcat_list[[i]] <- res$catalogue
}

# bind the catalogues in one table
SNV_catalogues <- do.call(cbind, SNVcat_list)
str(SNV_catalogues)
```


```{r}
# the catalogues can be plotted as follows
# plot_sum: show total vars in title
plotSubsSignatures(signature_data_matrix = SNV_catalogues,
                   plot_sum = TRUE, output_file = here("nogit/hrdetect/out/SNV_catalogues.jpg"))
```

```{r}
#fit the 12 breast cancer signatures using the bootstrap signature fit approach
sigsToUse <- c(1, 2, 3, 5, 6, 8, 13, 17, 18, 20, 26, 30)
subs_fit_res <- SignatureFit_withBootstrap_Analysis(
  outdir = here("nogit/hrdetect/out/sigFit/"),
  cat = SNV_catalogues,
  signature_data_matrix = COSMIC30_subs_signatures[,sigsToUse],
  type_of_mutations = "subs",
  nboot = 100, nparallel = 4)

#The signature exposures can be found here and correspond to the median
#of the boostrapped runs followed by false positive filters. See
#?SignatureFit_withBootstrap_Analysis for details
snv_exp <- subs_fit_res$E_median_filtered

```

```{r}
col_hrdetect <- c("del.mh.prop", "SNV3", "SV3", "SV5", "hrd", "SNV8")
input_matrix <- matrix(NA, nrow = length(sample_names),
                       ncol = length(col_hrdetect),
                       dimnames = list(sample_names, col_hrdetect))

# We have already quantified the amount of SNV signatures in the samples,
# so we can supply these via the input matrix
input_matrix[colnames(snv_exp),"SNV3"] <- snv_exp["Signature.3",]
input_matrix[colnames(snv_exp),"SNV8"] <- snv_exp["Signature.8",]

#run the HRDetect pipeline, for more information see ?HRDetect_pipeline
res <- HRDetect_pipeline(input_matrix,
                         genome.v = "hg19",
                         SV_bedpe_files = SV_bedpe_files,
                         Indels_vcf_files = Indels_vcf_files,
                         CNV_tab_files = CNV_tab_files,
                         nparallel = 2)

#save HRDetect scores
writeTable(res$hrdetect_output,file = here("nogit/hrdetect/out/HRDetect_res.tsv"))
```

## Single Sample

```{r}
snvindel_vcf <- system.file("extdata/umccrise/v0.18/snv/somatic-ensemble-PASS.vcf.gz", package = "gpgr")
snvindel_tabs <- hrdetect_read_snvindel_vcf(snvindel_vcf)

sv_vcf <- system.file("extdata/umccrise/v0.18/sv/manta.vcf.gz", package = "gpgr")
sv_bedpe <- hrdetect_read_sv_vcf(sv_vcf, nm = params$sample, hg_version = "hg38")

cnv_file <- system.file("extdata/purple/v2.39/purple.cnv.somatic.tsv", package = "gpgr")
cnv <- hrdetect_read_purple_cnv(cnv_file)

snv_catalogue <- signature.tools.lib::tabToSNVcatalogue(
  subs = snvindel_tabs$snv,
  genome.v = "hg38")
indel_classification <- signature.tools.lib::tabToIndelsClassification(
  indel.data = snvindel_tabs$indel,
  sampleID = params$sample,
  genome.v = "hg38")
sv_catalogue <- signature.tools.lib::bedpeToRearrCatalogue(sv_bedpe = sv_bedpe)
cnv_hrd <- signature.tools.lib::ascatToHRDLOH(ascat.data = cnv, SAMPLE.ID = params$sample)
```

