<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CHORD â€¢ gpgr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CHORD">
<meta property="og:description" content="gpgr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gpgr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/chord.html">CHORD</a>
    </li>
    <li>
      <a href="../articles/hrdetect.html">HRDetect</a>
    </li>
    <li>
      <a href="../articles/purple.html">PURPLE</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/umccr/gpgr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="chord_files/header-attrs-2.4/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CHORD</h1>
                        <h4 class="author">Peter Diakumis</h4>
            
            <h4 class="date">2020-10-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/umccr/gpgr/blob/master/vignettes/chord.Rmd"><code>vignettes/chord.Rmd</code></a></small>
      <div class="hidden name"><code>chord.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/umccr/gpgr">gpgr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/">devtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span></pre></div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>CHORD: Classifier of HOmologous Recombination Deficiency (<a href="https://github.com/UMCUGenetics/CHORD" class="uri">https://github.com/UMCUGenetics/CHORD</a>)</p>
<blockquote>
<p>CHORD is a random forest model that uses the relative counts of somatic mutation contexts to predict homologous recombination deficiency (HRD). The primary contexts used by CHORD are deletions with flanking microhomology and 1-100kb structural duplications. Additionally, 1-100kb structural duplications are used to distinguish BRCA1-type HRD from BRCA2-type HRD.</p>
</blockquote>
</div>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>CHORD takes as input somatic VCFs containing SNVs, INDELs and SVs. The former two can be in the same VCF. It is also able to take data.frames with those variants as input. Due to the annotation size in some of our processed Manta VCFs, we prefer to feed it a pre-processed SV data.frame instead of a VCF, thus we use the <code><a href="../reference/chord_mantavcf2df.html">gpgr::chord_mantavcf2df</a></code> function.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">snvindel_vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/umccrise/snv/somatic-ensemble-PASS.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"gpgr"</span><span class="op">)</span>
<span class="va">sv_vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/umccrise/sv/manta.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"gpgr"</span><span class="op">)</span></pre></div>
</div>
<div id="analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis" class="anchor"></a>Analysis</h2>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">sv_df</span> <span class="op">&lt;-</span> <span class="fu">gpgr</span><span class="fu">::</span><span class="fu"><a href="../reference/chord_mantavcf2df.html">chord_mantavcf2df</a></span><span class="op">(</span><span class="va">sv_vcf</span><span class="op">)</span> <span class="co"># prepare SV VCF as data.frame</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">gpgr</span><span class="fu">::</span><span class="fu"><a href="../reference/chord_run.html">chord_run</a></span><span class="op">(</span>
  vcf.snv <span class="op">=</span> <span class="va">snvindel_vcf</span>,
  df.sv <span class="op">=</span> <span class="va">sv_df</span>,
  sv.caller <span class="op">=</span> <span class="st">"manta"</span>,
  <span class="co"># vcf.sv = sv_vcf, # alternative</span>
  sample.name <span class="op">=</span> <span class="st">"sample_A"</span>,
  ref.genome <span class="op">=</span> <span class="st">"hg38"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; #====== Loading variants from vcfs ======#</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## SNVs</span>
<span class="co">#&gt; Warning in fun(libname, pkgname): </span>
<span class="co">#&gt;    No reference genome loaded. Please install and load a BSgenome.</span>
<span class="co">#&gt;    For example:</span>
<span class="co">#&gt;       install.packages('BiocManager')</span>
<span class="co">#&gt;       BiocManager::install('BSgenome.Hsapiens.UCSC.hg19')</span>
<span class="co">#&gt;       library('BSgenome.Hsapiens.UCSC.hg19')</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    Then specify the BSgenome to the ref.genome arguemnts to the relevant functions.</span>
<span class="co">#&gt;    For example:</span>
<span class="co">#&gt;       extractSigsSnv(..., ref.genome=BSgenome.Hsapiens.UCSC.hg19)</span>
<span class="co">#&gt; Reading in vcf file...</span>
<span class="co">#&gt; Converting chrom name style to style in ref.genome...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Indels</span>
<span class="co">#&gt; vcf file is the same for both SNVs and indels. Skipping reading vcf for indels</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## SVs</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; #====== Counting mutation contexts ======#</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Single base substitutions</span>
<span class="co">#&gt; Loading variants...</span>
<span class="co">#&gt; Removing rows with multiple ALT sequences...</span>
<span class="co">#&gt; Converting chrom name style to style in ref.genome...</span>
<span class="co">#&gt; Subsetting for SNVs...</span>
<span class="co">#&gt; Returning SNV trinucleotide contexts...</span>
<span class="co">#&gt; Initializing SNV signature output vector...</span>
<span class="co">#&gt; Converting trinucleotide contexts to substitution contexts...</span>
<span class="co">#&gt; Counting substitution context occurrences...</span>
<span class="co">#&gt; Returning context counts...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## Indel contexts (types x lengths)</span>
<span class="co">#&gt; Loading variants...</span>
<span class="co">#&gt; Removing rows with multiple ALT sequences...</span>
<span class="co">#&gt; Converting chrom name style to style in ref.genome...</span>
<span class="co">#&gt; Determining indel type...</span>
<span class="co">#&gt; Determining indel length and sequence...</span>
<span class="co">#&gt; Returning indel characteristics...</span>
<span class="co">#&gt; Initializing indel signature output vector...</span>
<span class="co">#&gt; Determining the start/end positions for the left/right flanks of each indel...</span>
<span class="co">#&gt; Retrieving flanking sequences...</span>
<span class="co">#&gt; Calculating the number of copies of the indel sequence are present in the 3' flanking sequence...</span>
<span class="co">#&gt; Calculating the (max) number of bases that are homologous to the 5'/3' flanking sequence...</span>
<span class="co">#&gt; Determining indel contexts...</span>
<span class="co">#&gt; Counting indel context occurrences...</span>
<span class="co">#&gt; Returning indel context counts...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ## SV contexts (types x lengths)</span>
<span class="co">#&gt; Creating SV type/length lookup table...</span>
<span class="co">#&gt; Counting DEL, DUP, and INV context occurrences...</span>
<span class="co">#&gt; Counting TRA occurrences...</span>
<span class="co">#&gt; Returning SV contexts...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; #====== Exporting output =========#</span>
<span class="co">#&gt; output.path not specified. Directly returning output</span>
<span class="co">#&gt; </span></pre></div>
</div>
<div id="results" class="section level2">
<h2 class="hasAnchor">
<a href="#results" class="anchor"></a>Results</h2>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">res</span>, list.len <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ contexts  : num [1, 1:145] 123 82 18 91 92 98 21 94 51 55 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr "sample_A"</span>
<span class="co">#&gt;   .. ..$ : chr [1:145] "A[C&gt;A]A" "A[C&gt;A]C" "A[C&gt;A]G" "A[C&gt;A]T" ...</span>
<span class="co">#&gt;  $ prediction:'data.frame':  1 obs. of  17 variables:</span>
<span class="co">#&gt;   ..$ sample           : chr "sample_A"</span>
<span class="co">#&gt;   ..$ p_BRCA1          : num 0.156</span>
<span class="co">#&gt;   ..$ p_BRCA2          : num 0.648</span>
<span class="co">#&gt;   ..$ p_hrd            : num 0.804</span>
<span class="co">#&gt;   .. [list output truncated]</span>

<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">prediction</span><span class="op">)</span>, value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">prediction</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>.name_repair <span class="op">=</span> <span class="st">"check_unique"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span>format <span class="op">=</span> <span class="st">"html"</span>, caption <span class="op">=</span> <span class="st">"CHORD results."</span><span class="op">)</span></pre></div>
<table class="table">
<caption>
CHORD results.
</caption>
<thead><tr>
<th style="text-align:left;">
var
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample
</td>
<td style="text-align:left;">
sample_A
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA1
</td>
<td style="text-align:left;">
0.156
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA2
</td>
<td style="text-align:left;">
0.648
</td>
</tr>
<tr>
<td style="text-align:left;">
p_hrd
</td>
<td style="text-align:left;">
0.804
</td>
</tr>
<tr>
<td style="text-align:left;">
hr_status
</td>
<td style="text-align:left;">
HR_deficient
</td>
</tr>
<tr>
<td style="text-align:left;">
hrd_type
</td>
<td style="text-align:left;">
BRCA2_type
</td>
</tr>
<tr>
<td style="text-align:left;">
remarks_hr_status
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
remarks_hrd_type
</td>
<td style="text-align:left;">
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA1.5%
</td>
<td style="text-align:left;">
0.0997
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA1.50%
</td>
<td style="text-align:left;">
0.178
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA1.95%
</td>
<td style="text-align:left;">
0.2939
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA2.5%
</td>
<td style="text-align:left;">
0.5171
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA2.50%
</td>
<td style="text-align:left;">
0.615
</td>
</tr>
<tr>
<td style="text-align:left;">
p_BRCA2.95%
</td>
<td style="text-align:left;">
0.6773
</td>
</tr>
<tr>
<td style="text-align:left;">
p_hrd.5%
</td>
<td style="text-align:left;">
0.746
</td>
</tr>
<tr>
<td style="text-align:left;">
p_hrd.50%
</td>
<td style="text-align:left;">
0.797
</td>
</tr>
<tr>
<td style="text-align:left;">
p_hrd.95%
</td>
<td style="text-align:left;">
0.8205
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit">

<span class="co"># grab first and last few of the contexts</span>
<span class="va">col_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">contexts</span><span class="op">)</span> <span class="op">-</span> <span class="fl">10</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">contexts</span><span class="op">)</span><span class="op">)</span>
<span class="va">res</span><span class="op">$</span><span class="va">contexts</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"context"</span>, .name_repair <span class="op">=</span> <span class="st">"check_unique"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/ranking.html">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">n</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="va">col_ind</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span>format <span class="op">=</span> <span class="st">"html"</span>, caption <span class="op">=</span> <span class="st">"Sample of rows from contexts counts."</span><span class="op">)</span></pre></div>
<table class="table">
<caption>
Sample of rows from contexts counts.
</caption>
<thead><tr>
<th style="text-align:right;">
n
</th>
<th style="text-align:left;">
context
</th>
<th style="text-align:right;">
sample_A
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
A[C&gt;A]A
</td>
<td style="text-align:right;">
123
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
A[C&gt;A]C
</td>
<td style="text-align:right;">
82
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
A[C&gt;A]G
</td>
<td style="text-align:right;">
18
</td>
</tr>
<tr>
<td style="text-align:right;">
4
</td>
<td style="text-align:left;">
A[C&gt;A]T
</td>
<td style="text-align:right;">
91
</td>
</tr>
<tr>
<td style="text-align:right;">
5
</td>
<td style="text-align:left;">
C[C&gt;A]A
</td>
<td style="text-align:right;">
92
</td>
</tr>
<tr>
<td style="text-align:right;">
6
</td>
<td style="text-align:left;">
C[C&gt;A]C
</td>
<td style="text-align:right;">
98
</td>
</tr>
<tr>
<td style="text-align:right;">
135
</td>
<td style="text-align:left;">
DUP_1e04_1e05_bp
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
136
</td>
<td style="text-align:left;">
DUP_1e05_1e06_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
137
</td>
<td style="text-align:left;">
DUP_1e06_1e07_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
138
</td>
<td style="text-align:left;">
DUP_1e07_Inf_bp
</td>
<td style="text-align:right;">
18
</td>
</tr>
<tr>
<td style="text-align:right;">
139
</td>
<td style="text-align:left;">
INV_0e00_1e03_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
140
</td>
<td style="text-align:left;">
INV_1e03_1e04_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
141
</td>
<td style="text-align:left;">
INV_1e04_1e05_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
142
</td>
<td style="text-align:left;">
INV_1e05_1e06_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
143
</td>
<td style="text-align:left;">
INV_1e06_1e07_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
144
</td>
<td style="text-align:left;">
INV_1e07_Inf_bp
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:right;">
145
</td>
<td style="text-align:left;">
TRA
</td>
<td style="text-align:right;">
77
</td>
</tr>
</tbody>
</table>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<table class="table">
<caption>
Platform information.
</caption>
<thead><tr>
<th style="text-align:left;">
var
</th>
<th style="text-align:left;">
value
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
version
</td>
<td style="text-align:left;">
R version 4.0.3 (2020-10-10)
</td>
</tr>
<tr>
<td style="text-align:left;">
os
</td>
<td style="text-align:left;">
Ubuntu 18.04.5 LTS
</td>
</tr>
<tr>
<td style="text-align:left;">
system
</td>
<td style="text-align:left;">
x86_64, linux-gnu
</td>
</tr>
<tr>
<td style="text-align:left;">
ui
</td>
<td style="text-align:left;">
X11
</td>
</tr>
<tr>
<td style="text-align:left;">
language
</td>
<td style="text-align:left;">
(EN)
</td>
</tr>
<tr>
<td style="text-align:left;">
collate
</td>
<td style="text-align:left;">
C.UTF-8
</td>
</tr>
<tr>
<td style="text-align:left;">
ctype
</td>
<td style="text-align:left;">
C.UTF-8
</td>
</tr>
<tr>
<td style="text-align:left;">
tz
</td>
<td style="text-align:left;">
Etc/UTC
</td>
</tr>
<tr>
<td style="text-align:left;">
date
</td>
<td style="text-align:left;">
2020-10-20
</td>
</tr>
</tbody>
</table>
<table class="table">
<caption>
Main packages used in this vignette.
</caption>
<thead><tr>
<th style="text-align:left;">
package
</th>
<th style="text-align:left;">
version
</th>
<th style="text-align:left;">
datestamp
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
base
</td>
<td style="text-align:left;">
4.0.3
</td>
<td style="text-align:left;">
2020-10-18
</td>
</tr>
<tr>
<td style="text-align:left;">
CHORD
</td>
<td style="text-align:left;">
2.0
</td>
<td style="text-align:left;">
2020-09-08
</td>
</tr>
<tr>
<td style="text-align:left;">
gpgr
</td>
<td style="text-align:left;">
0.0.11
</td>
<td style="text-align:left;">
2020-10-20
</td>
</tr>
<tr>
<td style="text-align:left;">
mutSigExtractor
</td>
<td style="text-align:left;">
1.13
</td>
<td style="text-align:left;">
2020-09-08
</td>
</tr>
</tbody>
</table>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Peter Diakumis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
